ethcore/src/blockchain/update.rs

```diff
@@ -22,3 +22,13 @@ pub struct ExtrasUpdate<'a> {
 	/// Modified transaction addresses (None signifies removed transactions).
 	pub transactions_addresses: HashMap<H256, Option<TransactionAddress>>,
 }
+
+/// Extra information in block insertion.
+pub struct ExtrasInsert {
+	/// Is the inserted block considered the best block.
+	pub is_new_best: bool,
+	/// Is the inserted block considered finalized.
+	pub is_finalized: bool,
```

Comment:
how can `is_new_best` be false and `is_finalized` be true? Any finalized block is always going to be in the fork-choice, and since we only insert fork tips an instantly finalized block would automatically be our new "best block".

Reply:
"finalized" means "stable in the fork-choice rule" (i.e. if block B is finalized, `fork_choice(B, B') = B for all B'`. How can we have a best block which is not included in the fork-choice? How can we have a block with no children, which is stable in the fork-choice rule, and call some other block the best block?

Reply:
> So a immediate finalized new block might not be the best block if it's score is less? 

This misses the entire point of casper. There is no purpose to implementing CasperFFG if we just ignore the finality information from it and continue to follow the chain with the highest total difficulty (a useless metric in the presence of finality; the only metric we would care about in that kind of hybrid system is TD-since-finalized-block).

Reply:
And in the above case, still consider `B` to be finalized already, and we insert `D`.

```
   B - C
  /
A
  \
   D
```

Even in the situation where `D` has `is_best_block` to be true (because for that we don't yet consulted the finalization rules), we'll never reorg from `C` to `D`. The actual best block will still be `C`.

Reply:
> If we use the finalized definition in the paper, this is theoretically allowed by CasperFFG 

This is not possible except with some supermajority of validators violating a slashing condition. We should ignore that for the time being and allow "de-finalization" and recovery processes for handling malicious validator sets in future work. For the purposes of the initial work here, let's only consider engines with a fault-tolerance assumption that doesn't permit safety failures.

> D has is_best_block to be true
 
`D` can't have `is_best_block` be true because `D` is no longer a member of the fork-choice. It's a contradiction. `D` might have the highest total difficulty when counting back to the genesis but it is not the best block.

> because we consider fork_choice and finalization rules separately

They shouldn't be considered separately because the fork choice depends on finality

Reply:
I think maybe we rename `is_best_block` to `is_best_block_pre_finalization`?

Check `is_best_block_pre_finalization` and `is_finalized` separately, and then reach a conclusion of the final `fork_choice` fits most of our use cases. The thing I worry if we move the whole finalization check to `Engine::fork_choice`, is that I think with that we also need to many reorg rule checks from `BlockChain` to `Engine`, which will make `Engine` big.

Reply:
> `D` can't have `is_best_block` be true because `D` is no longer a member of the fork-choice.

I agree. However, I don't think that should be in `Engine`. I would favor to put that logic to `BlockChain` or `Client` in the future once we implement "finalization pruning`, because if we know a block is outside of the finalization branch, we can discard it without checking any `Engine` rules.

Reply:
> is_best_block_pre_finalization

this is still inaccurate. it's more of a `is_best_block_assuming_nothing_has_ever_been_finalized` and I don't see how that's actually useful.

Reply:
`Engine::fork_choice(old_head, new_head) -> enum Choice { Left, Right }`.
and the blockchain makes sure no finalized blocks are reverted. Where does `is_best` even come into this?

Reply:
> the blockchain calls this and then makes sure no finalized blocks are reverted

Yeah that's the current behavior. At least I got that part right. :smile: 

* I still think that what might be confusing there is naming. So I'm going to rename `is_new_best` to `fork_choice`.
* Before we implement finalization pruning, we may still want to process a block even if it is confirmed to be on a branch. This avoids breaking too many code at once, as those blocks might still be needed somewhere. So I'm in favor of using the behavior "the blockchain calls `Engine::fork_choice` and then makes sure no finalized blocks are reverted".
* If I understand this correctly, the new block's finalization status `new.is_finalized` does not participate in the finalization check. Instead, it's the current best block's (and ancestry's) finalization status. We check whether reorg to the new block is blocked by finalization status of current best block (or its ancestry). The scope of blocks we can reorg to is set before the new block, and regardless of the new block's finalization status. So I think it's safe just to assume that it's an `Engine` logic failure (or really huge network attack) if we ever meet the cases discussed in this thread.

Reply:
> is_new_best to primitive_fork_choice

I don't see the purpose of that at all. What is the definition of a "primitive fork choice" and how do you derive it from a general fork choice function? Because it seems to me as just a way to introduce bug-prone complexity into the fork choice logic. What is the downside of removing it?

> the new block's finalization status new.is_finalized does not participate in the finalization check

it should, why wouldn't it? is it not part of the chain?

Reply:
> it should, why wouldn't it? is it not part of the chain?

Using the above example:

```
   B - C
  /
A
  \
   B' - C' - D' - E'
```

The reason why we don't reorg to `E'` is only because `B` is finalized. That is, we check `C` (current best block) and `C`'s ancestry finalization status. Yeah usually on `E'` branch there won't be any finalized block, given we don't have huge network attack. But the decisive factor is on `C`'s branch, not `E'`. Is my understanding on this point correct?

> What is the definition of a "primitive fork choice" and how do you derive it from a general fork choice function?

I think it's equivalent we write:

```
BlockChain::import_block(head, |old_head, new_head| MyEngine.fork_choice(old_head, new_head))
```

Or we write:

```
let primitive_fork_choice = MyEngine.fork_choice(old_head, new_head);
BlockChain::import_block(head, primitive_fork_choice);
```

If we combine the functions together, that would potentially mean we allow "finalization" to mean different things for different Engine. I just think that may not be a good idea. For Casper, `primitive_fork_choice` is basically Casper score comparison.

Another thing is, I don't think we can remove `primitive_fork_choice`. `fork_choice == primitive_fork_choice + finalization_rules` (basically as you said, "the blockchain calls this and then makes sure no finalized blocks are reverted"). But if we combine them together, I think we would probably need to move reorg logic to `Engine`, and my suggestion is that we try to break Engine trait before we make it too big..

Reply:
> The reason why we don't reorg to E' is only because B is finalized. That is, we check C (current best block) and C's ancestry finalization status. Yeah usually on E' branch there won't be any finalized block, given we don't have huge network attack. But the decisive factor is on C's branch, not E'. Is my understanding on this point correct?

This is correct but I don't see what it has to do with handling instant finality correctly. Attempting a general solution to handling of fault-tolerance violations (e.g. conflicting finalizations) is also probably going to be worse than doing nothing at all and as I suggest we should leave that for future work.

> let primitive_fork_choice = MyEngine.fork_choice(old_head, new_head);
It shouldn't be a "primitive" fork choice, it should be _the_ fork choice, or at least defined as a sub-decision after ensuring no finalized blocks would be reverted (which limits the class of fork-choice rules we can implement but not meaningfully until later-stage sharding).

There is a subtle but important difference between checking for reversion of finalized blocks before and after performing the remainder of the fork-choice operation. The former is more correct, but will behave the same as the latter in all cases except for malicious instant finality. Neither requires an `is_new_best` parameter.

Reply:
anyway, I guess the current solution is workable as long as it's well-documented that it is broken in that case. in the future we probably will need to move the reorg-checking stuff to engine anyway.

Reply:
As a concrete example of a fork-choice rule which doesn't fall into this category, imagine a situation where `Parity` nodes are running a parachain for Polkadot and they use a `PolkadotEngine` that wraps a polkadot node. The fork choice for `PolkadotEngine` would be to follow the most recently finalized polkadot block and not to follow the blockchain's notion of finality in any way.

Reply:
Okay got it. Yeah indeed it won't cover that instant finalization case..

Anyway, I'll add more documentations on `Engine::fork_choice` for how that's currently working.